### DISCUSSION TITLE
Persistent Redis сonnections in Sidekiq with Async::Redis

### DISCUSSION THREAD

**Discussion by Alexey2257** [ioquatix/ioquatix#34] [General]
Hi! I am a big fan of Async and try to bring it into our project as much as possible. However, there are some things that are not obvious at first sight, like the need for a reactor for persistent connections in gems like Async::Redis. In our project with Rails and Sidekiq, I came up with a monkey-patch for Sidekiq and wrote an article about it. I think it might be useful for people who are starting to use the Async library like me.  I would love to hear some feedback: is my approach correct?

https://dev.to/alexey2257/persistent-redis-connections-in-sidekiq-with-asyncredis-a-deep-dive-f3c

**Comment by ioquatix** [comment]
I haven't finished reading the article yet, but so far it looks good.

My first feedback is to check https://socketry.github.io/async-pool/guides/getting-started/index#limit-and-concurrency

> Note that you should pass either limit or concurrency if you want to have more than one connection because the default in Async::Pool is 1 (and concurrency == limit by default):

By default, the connection pool has no limit (so it will create an unbounded number of resources/connections), but the concurrency is set to 1 which means it will create at most one resource at a time, no matter how many times it gets acquired.

**Reply by Alexey2257** [reply]
Thank you! My bad for not checking the doc, I relied on watching source code and got it wrong. I've changed this part and also added a link to the doc.

**Comment by ioquatix** [comment]
Okay, I finished reading the article, actually I really like it. It shows the perfect way to integrate Async with Sidekiq. It's also neat that the actual integration is so simple.

I think one challenge of this approach is that you are inherently limited by the job scheduling of Sidekiq's sequential loop. In other words, you are getting the benefits of event driven concurrency and connection pooling, but it appears limited to sequential processing of jobs on each reactor. In addition, having multiple reactors per thread while okay, will encounter GVL contention which can become a big problem as you start to scale up using threads.

You can solve the GVL issue by only running one thread per worker. However that obviously limits your scalability - and there can be some benefits to using multiple threads (and also other downsides like 3rd party code instability).

In order to get concurrency back in such a scenario, it must be possible to schedule several jobs at a time:

```ruby
    def run
      # By setting this thread-local, Sidekiq.redis will access +Sidekiq::Capsule#redis_pool+
      # instead of the global pool in +Sidekiq::Config#redis_pool+.
      Thread.current[:sidekiq_capsule] = @capsule

      Sync do |task|
        idler = Async::Idler.new

        while !@done
          idler.wait
          task.async{process_one}
        end
      end

      @callback.call(self)
    rescue Sidekiq::Shutdown
      @callback.call(self)
    rescue Exception => ex
      @callback.call(self, ex)
    end
```

The above (untested) code uses an idler which will wait if the system is getting overloaded (80% utilization target by default) and then schedule a job using `async{process_one}`. This will schedule multiple jobs in the event loop. However, it might break the error handling and/or other assumptions about how `def run` works... cc @mperham

This is essentially how my own `async-job` job server works: https://github.com/socketry/async-job/blob/bf7d806b8a06a857bbdd07cc81e4367807203d6c/lib/async/job/backend/redis/server.rb#L48

**Reply by Alexey2257** [reply]
Thanks for the feedback! I agree that indeed my approach has it's limitations due to Sidekiq threaded nature. This actually was done for an existing project, where I found out that using async inside of some jobs gives performance improvement already, so it's 20/80. I thought about using capsules for limitation to 1 thread for jobs with async but this leads for even more thread-concurrency if sidekiq worker has other jobs to do. My next step will be to try async-job in this project, I will reach you out as soon as I get some results with it! By the way, here is the result we already reached with Async, 2x less time spent for the job:
![изображение](https://github.com/user-attachments/assets/c2825fbc-2720-479f-89fb-ecde678f7341)

**Reply by ioquatix** [reply]
What do you think is giving the performance improvement? Persistent connections? It's not only looking better, but less variance too.

**Reply by Alexey2257** [reply]
No, persistent connections were just a side quest I dealt with after using async here, so I decided to share it in the article.
As for the performance:
We used to have such a worker
``` ruby
def perform(params)
  DoSomeWork.call(*params)
end
```
where DoSomeWork is I/O intensive: it fetches data from an API and then saves it to Postgres and Redis. It's actually a bit more complicated and includes a code mess, so, for example, you can't easily extract the part that fetches data from the API.
For this worker, Sidekiq default concurrency was not enough (and we are not able to increase it because we have other workers), so I came up with a solution to do threads inside a thread:
``` ruby
  def perform(grouped_params)
    futures = grouped_params.map do |params|
      Concurrent::Future.execute { DoSomeWork.call(*params) }
    end

    futures.each(&:wait)
  end
```
That gave us about a 30% performance improvement. Finally, after Rails improved support for fibers (introduction of the isolation level), I could do a no-brain replacement with Async:
``` Ruby
  def perform(grouped_params)
    grouped_params.map do |params|
      Async do
         DoSomeWorkcall(*params)
      end
    end.each(&:wait)
  end
```
And that gave additional 50%. I think we can achieve better if extract "fetch from api" part from service because we could do more concurrent http calls than concurrent Postgres requests and that's current bottleneck of `grouped_params` size. But even more interesting here would be to have some job processing based on Async instead of Sidekiq

**Comment by mperham** [comment]
I don't mind minor changes to make Sidekiq work better with Async but I don't see any significant push for async from the user community so it's not something I feel the need to officially support.

**Reply by ioquatix** [reply]
That's reasonable, more just FYI.
